### DISCUSSION TITLE
Under which conditions can `GlobSetBuilder::build()` error?

### DISCUSSION THREAD

**Discussion by konstin** [BurntSushi/ripgrep#2927] [Q&A] [answered]
`GlobSetBuilder::build()` returns a `Result<GlobSet, Error>`, even though it's constructed from valid globs. Under which conditions does it raise the error?

**Comment by BurntSushi** [comment, marked as answer]
This variant: https://github.com/BurntSushi/ripgrep/blob/4b0e5a11b9c5a35fc4553c527051489c5f97ece0/crates/globset/src/lib.rs#L177-L178

Specifically, from here: https://github.com/BurntSushi/ripgrep/blob/4b0e5a11b9c5a35fc4553c527051489c5f97ece0/crates/globset/src/lib.rs#L292-L295

Something to note here is that a `Glob` itself can't do any matching. Basically, a `Glob` is just the result of converting a glob pattern to a regex. That can fail because of invalid glob syntax. But the process of converting a `Glob` to a `GlobMatcher` (for matching one glob) or many `Glob`s to a `GlobSet` (for matching zero or more globs) can still fail. I believe the only possible error here is if the size limit is exceeded. That size limit is currently hard-coded:

https://github.com/BurntSushi/ripgrep/blob/4b0e5a11b9c5a35fc4553c527051489c5f97ece0/crates/globset/src/lib.rs#L286

So basically, if your globs combine into one compiled regex that is too big, then this step could fail.

Since Unicode handling is generally disabled in this crate, and since the glob syntax doesn't give you access to things that would otherwise make it easy to build huge regexes (like bounded repetitions), you would likely need many globs to blow the size limit.

**Reply by konstin** [reply]
Thanks, I'll add an error variant as "glob expressions caused to large regex".
